crates/rbuilder/src/building/evm_inspector.rs

```diff
@@ -104,9 +128,65 @@ where
                         .insert(key, written_value);
                 }
             }
+            opcode::BALANCE => {
+                if let Ok(addr) = interpreter.stack().peek(0) {
+                    let addr = Address::from_word(B256::from(addr.to_be_bytes()));
+                    self.next_step_action = NextStepAction::ReadBalanceResult(addr);
+                }
+            }
+            opcode::SELFBALANCE => {
+                let addr = interpreter.contract().target_address;
+                self.next_step_action = NextStepAction::ReadBalanceResult(addr);
+            }
             _ => (),
         }
     }
+
+    fn call(&mut self, _: &mut EvmContext<DB>, inputs: &mut CallInputs) -> Option<CallOutcome> {
+        if let Some(transfer_value) = inputs.transfer_value() {
+            if !transfer_value.is_zero() {
+                *self
+                    .used_state_trace
+                    .sent_amount
+                    .entry(inputs.transfer_from())
+                    .or_default() += transfer_value;
+                *self
+                    .used_state_trace
+                    .received_amount
+                    .entry(inputs.transfer_to())
+                    .or_default() += transfer_value;
+            }
+        }
+        None
+    }
+
+    fn create_end(
+        &mut self,
+        _: &mut EvmContext<DB>,
+        _: &revm::interpreter::CreateInputs,
+        outcome: revm::interpreter::CreateOutcome,
+    ) -> revm::interpreter::CreateOutcome {
+        if let Some(addr) = outcome.address {
+            self.used_state_trace.created_contracts.push(addr);
+        }
+        outcome
+    }
+
+    fn selfdestruct(&mut self, contract: Address, target: Address, value: U256) {
+        self.used_state_trace.destructed_contracts.push(contract);
```

Comment:
This should check if contract was already destroyed and return if yes. Selfdestruct can be called multiple times in one tx.